tensorflow/contrib/rnn/kernels/lstm_ops.cc

```diff
@@ -0,0 +1,599 @@
+                                        b_tensor->dim_size(0),
+                                        " vs. ", cell_size_ * 4));
+
+    perftools::gputools::Stream* stream =
+        ctx->op_device_context() ? ctx->op_device_context()->stream() : nullptr;
+
+    OpOutputList x_grad_tensors;
+    OP_REQUIRES_OK(ctx, ctx->output_list("x_grad", &x_grad_tensors));
+
+    Tensor* w_grad_tensor = nullptr;
+    OP_REQUIRES_OK(ctx, ctx->allocate_output("w_grad",
+          TensorShape({input_size + cell_size_, cell_size_ * 4}),
+          &w_grad_tensor));
+    TensorMemZero(w_grad_tensor, stream);
+
+    Tensor* b_grad_tensor = nullptr;
+    OP_REQUIRES_OK(ctx, ctx->allocate_output("b_grad",
+          TensorShape({cell_size_ * 4}), &b_grad_tensor));
+    TensorMemZero(b_grad_tensor, stream);
+
+    for (int64 t = 0; t < sequence_len_max_; ++t) {
+      Tensor* x_grad_tensor = nullptr;
+      x_grad_tensors.allocate(
+          t, TensorShape({batch_size, input_size}), &x_grad_tensor);
+      TensorMemZero(x_grad_tensor, stream);
+    }
+
+    Tensor xh_tensor;
+    OP_REQUIRES_OK(ctx, ctx->allocate_temp(
+          DT_FLOAT, TensorShape({batch_size, input_size + cell_size_}),
+          &xh_tensor));
+
+    Tensor xh_grad_tensor;
+    OP_REQUIRES_OK(ctx, ctx->allocate_temp(DT_FLOAT,
+        TensorShape({batch_size, input_size + cell_size_}), &xh_grad_tensor));
+
+    Tensor states_grad_tensor;
+    OP_REQUIRES_OK(ctx, ctx->allocate_temp(DT_FLOAT,
+        TensorShape({batch_size, state_size}), &states_grad_tensor));
+    TensorMemZero(&states_grad_tensor, stream);
+
+    Tensor states_prev_grad_tensor;
+    OP_REQUIRES_OK(ctx, ctx->allocate_temp(DT_FLOAT,
+        TensorShape({batch_size, state_size}), &states_prev_grad_tensor));
+
+    for (int64 t = sequence_len_max - 1; t >= 0; --t) {
+      const Tensor& x_tensor = x_tensors[t];
+      const Tensor& states_prev_tensor =
+          t <= 0 ? *initial_state_tensor : states_tensors[t - 1];
+      const Tensor& states_tensor = states_tensors[t];
+      const Tensor& h_grad_tensor = h_grad_tensors[t];
+
+      Tensor* x_grad_tensor = x_grad_tensors[t];
+      const Tensor& states_grad_const_tensor = states_grad_tensor;
+
+      functor::LSTMCellBlockBprop<Device, USE_CUBLAS>()(
+          stream, ctx->eigen_device<Device>(),
+          batch_size, input_size, cell_size_,
+          x_tensor.matrix<float>(),
+          xh_tensor.matrix<float>(),
+          states_prev_tensor.matrix<float>(),
+          w_tensor->matrix<float>(),
+          b_tensor->vec<float>(),
+          states_tensor.matrix<float>(),
+          h_grad_tensor.matrix<float>(),
+          states_grad_const_tensor.matrix<float>(),
+          xh_grad_tensor.matrix<float>(),
+          x_grad_tensor->matrix<float>(),
+          states_prev_grad_tensor.matrix<float>(),
+          w_grad_tensor->matrix<float>(),
+          b_grad_tensor->vec<float>());
+
+      if (stream) {
```

Comment:
I think if you make this code part of the functor above, you won't have to check stream / use ifdef guards, since the device template  specializations will allow you to split your GPU and CPU code separately